### PR TITLE
Op join

### PR DESCRIPTION
**Pull request by swilly22**
Add support for JOIN clause

### CODE REVIEW

File: src/ast/ast_validations.c [hunk 1]
```diff
@@ -1409,6 +1400,83 @@ static AST_Validation _ValidateClauses(const AST *ast, char **reason) {
...
+
+	const cypher_astnode_t *return_clause = cypher_ast_query_get_clause(ast->root, return_indices[0]);
+	uint proj_count = cypher_ast_return_nprojections(return_clause);
+	const char *projections[proj_count];
+
+	for(uint j = 0; j < proj_count; j++) {
+		const cypher_astnode_t *proj = cypher_ast_return_get_projection(return_clause, j);
+		const cypher_astnode_t *alias_node = cypher_ast_projection_get_alias(proj);
+		if(alias_node == NULL)  {
+			projections[j] = NULL;
```
Comment (DvirDukhan) [inline src/ast/ast_validations.c]: In which cases projection will have NULL alias?

File: src/execution_plan/ops/op_join.c [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+static Record JoinConsume(OpBase *opBase) {
+	OpJoin *op = (OpJoin *)opBase;
+	Record r = NULL;
+
+	while(true) {
+		// Try pulling from current stream.
+		r = OpBase_Consume(op->stream);
+		// Managed to get a record, break out.
+		if(r) break;
```
Comment (DvirDukhan) [inline src/execution_plan/ops/op_join.c]: `if(r) return r;` 
I think it is cleaner.

File: src/execution_plan/ops/op_join.c [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+		// Stream depleted, see if there's a new stream to pull from.
+		op->streamIdx++;
+		if(op->streamIdx < op->op.childCount) {
+			op->stream = op->op.children[op->streamIdx];
+			// Re-enter loop.
+			continue;
+		}
+		break;
```
Comment (DvirDukhan) [inline src/execution_plan/ops/op_join.c]: `return NULL;`
I think it is cleaner.

File: src/execution_plan/execution_plan.c [hunk 4]
```diff
@@ -725,7 +814,8 @@ ExecutionPlan *NewExecutionPlan(RedisModuleCtx *ctx, GraphContext *gc, ResultSet
 	// Optimize the operations in the ExecutionPlan.
 	optimizePlan(gc, plan);
 
-	plan->segment_count = segment_count;
+	// Disregard self.
+	plan->segment_count = segment_count - 1;
```
Comment (DvirDukhan) [inline src/execution_plan/execution_plan.c]: can you explain?

Comment (swilly22) [reply src/execution_plan/execution_plan.c]: the last computed segment is acting as the complete execution-plan, 
plan = segments[segment_count-1]
there's no need for plan's segment to contain a reference to itself.

File: src/execution_plan/ops/op_join.c (hunk 2 above)
Comment (swilly22) [reply src/execution_plan/ops/op_join.c]: Restructured.

File: src/execution_plan/ops/op_join.c (hunk 3 above)
Comment (swilly22) [reply src/execution_plan/ops/op_join.c]: Restructured.

File: src/ast/ast_validations.c (hunk 1 above)
Comment (swilly22) [reply src/ast/ast_validations.c]: `RETURN a`
VS
`RETURN a AS b`

**Pull request merged**